DOCSP-23431 Continuous defragmentation edits (Flutter, Swift) (#2620)
mongodb/docs-realm: source/sdk/flutter/realm-database/realm-files/compact.txt

   :depth: 2
   :class: singlecol

The size of a Realm Database file is always larger than the total size of 
the objects stored within it. This architecture enables some of realm's 
performance, concurrency, and safety benefits.

Realm writes new data within unused space tracked inside a
file. In some situations, unused space may comprise a significant 
portion of a realm file. Realm's default behavior is to automatically 
compact a Realm database file to prevent it from growing too large. 
You can use manual compaction strategies when 
automatic compaction is not sufficient for your use case.

Automatic Compaction
--------------------

.. versionadded:: 0.9.0

.. include:: /includes/automatic-compaction.rst 

Manual Compaction Strategies

> ----------------------------

Reviewer: why doesn't this section use the include `manual-compaction-note.rst`?

Author: The note included the portion about not using automatic compaction if you "use an older version of the SDK that does not support automatic compaction". This doesn't apply to Flutter since it implemented Automatic pre-GA, but should apply to the other SDKs.

If automatic compaction is deemed insufficient, manual compaction 
can be used for applications that require stricter management 
of file size to improve performance. A production application should 

> implement manual compaction to periodically reduce the realm file size.

Reviewer: > A production application should 
implement manual compaction to periodically reduce the realm file size.

i believe this is not true. i think it was only nec before automatic compaction was a thing.

   :depth: 2
   :class: singlecol

The size of a Realm Database file is always larger than the total size of 
the objects stored within it. This architecture enables some of realm's 
performance, concurrency, and safety benefits.

Realm writes new data within unused space tracked inside a
file. In some situations, unused space may comprise a significant 
portion of a realm file. Realm's default behavior is to automatically 
compact a Realm database file to prevent it from growing too large. 
You can use manual compaction strategies when 
automatic compaction is not sufficient for your use case.

Automatic Compaction
--------------------

.. versionadded:: 0.9.0

.. include:: /includes/automatic-compaction.rst 

Manual Compaction Strategies
----------------------------

If automatic compaction is deemed insufficient, manual compaction 
can be used for applications that require stricter management 
of file size to improve performance. A production application should 
implement manual compaction to periodically reduce the realm file size

> if automatic compaction is not used.

Reviewer: active voice

```suggestion
if it does not use automatic compaction.
```

compact when you reach a certain file size or when the file size negatively 
impacts performance.

Automatic Compaction
--------------------

.. versionadded:: 0.9.0

The SDK automatically compacts Realm files in the background by continuously reallocating data 
and removing unused file space. Automatic compaction only takes place when
the file is not being accessed. 

Manual Compaction Strategies
----------------------------

> In most circumstances, automatic compaction will be sufficient for minimizing

Reviewer: ```suggestion
In most circumstances, automatic compaction is sufficient for minimizing
```

the Realm file size. Manual compaction can be used if automatic compaction does not meet 

> application needs, such as if your SDK does not support automatic compaction.

Reviewer: ```suggestion
application needs or if you're using an older version of the SDK
that does not support automatic compaction.
```

compact when you reach a certain file size or when the file size negatively 
impacts performance.

Automatic Compaction
--------------------

As of Flutter version 0.9.0 and up, Realm files are automatically compacted in the 

> background by continuously reallocating data in the file and removing 

Reviewer: > Realm files are automatically compacted in the background

please rephrase to active voice. perhaps smthn like: "the SDK automatically compacts Realm files in the background"

> unused file space. Files are automatically compacted once free space is twice the size

Reviewer: > As of Flutter version 0.9.0 and up, Realm files are automatically compacted in the background by continuously reallocating data in the file and removing 
unused file space.

i recommend splitting this idea into 2 sentences. the second part "by continuously reallocating data in the file and removing unused file space" is an implementation detail that most users probably don't care about. perhaps putting this idea in a second paragraph, focusing the first paragraph on just the behavior (which is all that's relevant for most all users).

Author: Made slightly shorter with active voice, is the sentence more digestible now?

> of Realm user data in the file. Automatic compaction will only take place when
> the file is not being accessed. 

Reviewer: ```suggestion
of Realm user data in the file. Automatic compaction only takes place when
the file is not being accessed. 
```

  environment. This is a property on realm's :flutter-sdk:`LocalConfiguration 
  <realm/LocalConfiguration-class.html>`, and only works with non-Synced realms.

.. note:: When To Use Manual Compaction

   Automatic compaction begins once free space in the file is twice the 
   size of user data in the file. Use the following manual compaction strategies 

>    if this automatic compaction threshold is not sufficient for your application. 

Reviewer: i don't think this should be in a note. it's a pretty important idea to understanding why to use manual compaction. i suggest putting this as the first paragraph

Reviewer: > Automatic compaction begins once free space in the file is twice the size of user data in the file.

i don't understand what this means. can you clarify?
